### ISSUE TITLE
[Issue] Version 2.3.4 created a type error with the `createMock` function

### ISSUE THREAD

**Issue by C0ZEN** [Typescript-TDD/ts-auto-mock#464]
### Subject of the issue
The latest release [2.3.4](https://github.com/Typescript-TDD/ts-auto-mock/releases/tag/2.3.4) related to the issue #458 fixed an error and also created a new one.

### Your environment
* ts-auto-mock version: 2.3.4
* typescript version: 3.9.7
* node version: 13.7.0
* npm version: 6.14.6

### Steps to reproduce

**Test:**

```typescript
client = createMock<Client>({
   user: {
      setPresence: setPresenceMock,
   },
});
```

The new error is on `user`

**Error:**
```typescript
TS2322: Type '{ setPresence: jest.Mock<Promise<Presence>, any>; }' is not assignable to type 'PartialObjectDeep<ClientUser>'.
Types of property 'valueOf' are incompatible.
Type '() => Object' is not assignable to type '(() => string) & (() => Object)'.
Type '() => Object' is not assignable to type '() => string'.
Type 'Object' is not assignable to type 'string'.
```

**Fake Client class:**

```typescript
export class Client {
    public user: ClientUser | null;
}
```

**Fake ClientUser class:**

```typescript
export class ClientUser {
    public setPresence(data: PresenceData): Promise<Presence>;
}
```

### Expected behavior
No error.

### Actual behavior
Type error.

**Comment by uittorio** [comment]
hi @C0ZEN, could you share the type of setPresenceMock?

**Comment by C0ZEN** [comment]
@uittorio if you need all these data I think that the best solution is to give you a complete reproduction.
Note that all types are coming from firebase-admin package though.

**Comment by uittorio** [comment]
I'll just install firebase-admin then!! thanks :P

**Comment by uittorio** [comment]
I've tried to install firebase-admin, but I couldn't find those types.  Sorry to bother you again, I just need the imports references of those. Thanks

**Comment by C0ZEN** [comment]
@uittorio oh sorry I meant discord.js

**Comment by C0ZEN** [comment]
@uittorio the reproduction is available: https://github.com/C0ZEN/ts-auto-mock-issue-464

**Comment by uittorio** [comment]
Cool, thanks @C0ZEN.

The issue is related to the default implementation in an object of valueOf. 

The deepPartial that has been added in 2.2.4 is much smarter and it does consider functions. `objects` have by default valueOf.

I'm not sure yet 100% if there is a way to workaround that in the deepPartial. I'll let you know.

In the meantime I would recommend to cast the type.

**Comment by C0ZEN** [comment]
Workaround for now.

```typescript
client = createMock<Client>({
	user: {
		setPresence: setPresenceMock,
	} as unknown as ClientUser,
});
```

**Comment by C0ZEN** [comment]
@uittorio I hope what you will be able to fix it because this new error breaks a lot of existing tests and is more painful than the previous error before the release.
I will still force the casting and mark all the tests with a todo.
Moreover this new version could eventually breaks other projects too 😢

**Comment by uittorio** [comment]
Yeah, I agree! I didn't realise that this change may impact a lot of scenario, I thought it was just a better type :).

I'll make the Partial Deep more relaxed to avoid these scenarios until we've found a better solution

**Comment by uittorio** [comment]
@C0ZEN . I've reproduce the issue just using Partial from typescript

```typescript
interface A {
    valueOf(): number;
    b: string;
  }

/*
Error TS2322: Type '{ b: string; }' is not assignable to type 'Partial<A>'.   The types returned by 'valueOf()' are incompatible between these types.     Type 'Object' is not assignable to type 'number'.
*/
  const a: Partial<A> = {
    b: 'hello 3',
  };


```

**Comment by C0ZEN** [comment]
@uittorio I am glad to see that with such a frequently type the error can be reproduced.
I hope that you will come with a solution 🙏🏻

**Comment by uittorio** [comment]
In my opinion the best approach, for now, its to re implement valueOf. That would also solve the Infinite issue that you are facing and you can remove the casting. 

```typescript

// first solution
  const mock: Client = createMock<Client>({
    user: {
      setPresence: jest.fn(),
      valueOf: jest.fn(),
    },
  });

  // second solution
  const mock2: Client = createMock<Client>();

  (mock2.user as ClientUser).setPresence = jest.fn();
```

**Comment by C0ZEN** [comment]
@uittorio ok it seems to me that indeed if both issues can be solved with creating the `valueOf` this is the best course of action.
Do you think that this is a good idea to open an issue in discord.js about the type error of the `valueOf`? They could maybe find a solution to change it and avoid this. Long shot but I can open an issue we never know.

Also even if there is a workaround could you please not close these issues?

**Comment by uittorio** [comment]
I don't think it's an issue on Discord js. They are entitled to override valueOf. I'll keep these issues open. We'll try and find a better approach :) Thanks for investigating and reporting those issues.

**Comment by Pmyl** [comment]
I realised that the problem is in how javascript works. When you create a new object with `{}` you're actially creating an object with prototype Object, meaning that in the type you also have methods like `toString` and `valueOf`.

This means that when you write:
```ts
client = createMock<Client>({
   user: {
      setPresence: setPresenceMock,
   },
});
```
in terms of typings is the same as writing:
```ts
client = createMock<Client>({
  user: {
    setPresence: setPresenceMock,
    valueOf: function() { ... },
    toString: function() { ... },
    ...
  },
});
```

Object is defined like this:
```ts
interface Object {
  ...

  /** Returns the primitive value of the specified object. */
  valueOf(): Object;

  ...
}
```

so this happens:
```ts
// type of 
{
  setPresence: setPresenceMock,
}
// is
{
  setPresence: Function,
  valueOf: () => Object,
  ...
}
```

The type of `valueOf` doesn't match the type of `valueOf` defined in ClientUser (that is `valueOf: () => string`) and the typings are failing.


All of this to say that the error makes sense. I think the only thing we can do about this is updating PartialDeep to be relaxed when using Object. In my mind we should change the name of the type and make it specific for the override parameter (something like MockOverrideValues. but better).
We know that if the property is part of the prototype it doesn't get copied over:
```_.merge({ valueOf: function() { return "123"; }}, {}).valueOf() === "123"```
this means that in PartialDeep when we get to the Object case we should strip it of its prototype properties fixing this bug.

Let me know if I made any mistakes
